### PR TITLE
Accept integers and floats in most functions

### PR DESCRIPTION
**Pull request by wojtekmach**

### CODE REVIEW

**Comment by ericmj** [comment]
I'm generally 👍 on this, but it might be dangerous to do automatic float conversion since we use `:io_lib_format.fwrite_g/1`. Is this is a 100% exact conversion?

/cc @josevalim

**Comment by josevalim** [comment]
@ericmj you may consider also using `Float.ratio`. `:io_lib_format.fwrite_g` returns the smallest number that will be represented by that float but there may be loss of precision. I am unsure.

**Comment by wojtekmach** [comment]
@ericmj yup, that's a valid concern. From what I could tell, `:io_lib_format.fwrite_g` (which `Float.to_string/1` uses nowadays too) is supposed to be accurate but I haven't yet read neither paper nor implementation.

I think it'd be awkward to accept integers but not accept floats, especially that Decimal.new of course accepts both. But I totally get that if there's uncertainty it would be better to have user do `Decimal.new(some_float)` first.

@josevalim good point about `Float.ratio`. We'd need to bump decimal to depend on Elixir 1.4, or vendor this function.

**Comment by wojtekmach** [comment]
@ericmj I just realized that because I'm basically delegating to `Decimal.new` in the changed functions, this implementation allows to also do e.g.:

```elixir
Decimal.add(0, "1.0")
#Decimal<1.0>
```

which on one hand is kinda neat because it's exact but on the other hand it "taints" all these functions with the fact that converting the string to decimal might raise. Is handling strings here desirable?

Maybe it'd be better to just ship automatic integer conversions initially and then add floats and/or strings?

**Comment by ericmj** [comment]
@wojtekmach What's your use cases for converting floats to Decimals? For me it feels wrong to do some work with floats and then later convert them to decimals. If you need literal floating decimals it's better to use strings: `Decimal.new("3.14")`. Maybe we should introduce a sigil: `~D[3.14]` to make this easier?

Relying on float literals will be extra bad if we change the behaviour by using `Float.ratio` instead, because `Decimal.new(3.14)` will change depending on the precision in your context since it now also involves a division.

File: lib/decimal.ex [hunk 1]
```diff
@@ -289,6 +299,10 @@ defmodule Decimal do
     context(%Decimal{sign: sign, coef: Kernel.abs(coef), exp: exp})
   end
 
+  def add(num1, num2) do
```
Comment (wojtekmach) [inline lib/decimal.ex]: @ericmj ok, let's extend this function to just accept integers. Do we want something like this:

```elixir
  def add(%Decimal{} = num1, num2) when is_integer(num2), do: add(num1, new(num2))

  def add(num1, %Decimal{} = num2) when is_integer(num1), do: add(new(num1), num2)

  def add(num1, num2) when is_integer(num1) and is_integer(num2), do: new(num1 + num2)
```

or there's a better way? The current implementation that just uses new/1 would also allow to pass floats and strings.

**Comment by wojtekmach** [comment]
Closing in favor of #71

**Pull request closed without merging**